direct/core/src/main/java/cz/o2/proxima/direct/batch/TerminationContext.java

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright 2017-2020 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.direct.batch;
+
+import cz.o2.proxima.annotations.Internal;
+import cz.o2.proxima.util.ExceptionUtils;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+
+/** A context that guards correct termination of batch observation process. */
+@Internal
+@Slf4j
+public class TerminationContext {
+
+  private final BatchLogObserver observer;
+  private final CountDownLatch terminateLatch = new CountDownLatch(1);
+  private volatile boolean cancelled = false;
+  private volatile Thread runningThread;
+
+  public TerminationContext(BatchLogObserver observer) {
+    this.observer = observer;
+  }
+
+  public boolean isCancelled() {
+    return cancelled || runningThread == Thread.currentThread() && runningThread.isInterrupted();
```

Comment:
can you elaborate little bit about why it not enough to check only `cancelled` flag here?

Reply:
Some runners (e.g. flink) might interrupt the running thread to terminate it.